server/auth/middleware.go

```diff
@@ -0,0 +1,257 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/bitcoin-sv/spv-wallet/config"
+	"github.com/bitcoin-sv/spv-wallet/engine"
+	"github.com/bitcoin-sv/spv-wallet/engine/utils"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/bitcoinschema/go-bitcoin/v2"
+	"github.com/gin-gonic/gin"
+	"github.com/libsv/go-bt/v2/bscript"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	// ParamXPubKey the request parameter for the xpub string
+	ParamXPubKey = "xpub"
+
+	// ParamXPubHashKey the request parameter for the xpub ID
+	ParamXPubHashKey = "xpub_hash"
+
+	// ParamAccessKey the request parameter for the xpub ID
+	ParamAccessKey = "access_key"
+
+	// ParamAdminRequest the request parameter whether this is an admin request
+	ParamAdminRequest = "auth_admin"
+
+	// ParamAuthSigned the request parameter that says whether the request was signed
+	ParamAuthSigned = "auth_signed"
+)
+
+// AuthPayload is the authentication payload for checking or creating a signature
+type AuthPayload struct {
+	AuthHash     string `json:"auth_hash"`
+	AuthNonce    string `json:"auth_nonce"`
+	AuthTime     int64  `json:"auth_time"`
+	BodyContents string `json:"body_contents"`
+	Signature    string `json:"signature"`
+	xPub         string
+	accessKey    string
+}
+
+// CorsMiddleware is a middleware that handles CORS.
+func CorsMiddleware() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		c.Writer.Header().Set("Access-Control-Allow-Origin", "*")
+		c.Writer.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
+		c.Writer.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization, Cache-Control")
+		c.Writer.Header().Set("Access-Control-Allow-Credentials", "true")
+		c.Next()
+	}
+}
+
+// AuthMiddleware will check the request for the xPub or AccessKey header
+func AuthMiddleware(engine engine.ClientInterface, appConfig *config.AppConfig) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		xPub := strings.TrimSpace(c.GetHeader(models.AuthHeader))
+		authAccessKey := strings.TrimSpace(c.GetHeader(models.AuthAccessKey))
+		if len(xPub) == 0 && len(authAccessKey) == 0 {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "missing authentication header"})
+			return
+		}
+
+		fmt.Println("xPub: ", xPub)
+		fmt.Println("authAccessKey: ", authAccessKey)
+
+		xPubID := utils.Hash(xPub)
+		xPubOrAccessKey := xPub
+
+		if xPub != "" {
+			// Validate that the xPub is an HD key (length, validation)
+			if _, err := utils.ValidateXPub(xPubOrAccessKey); err != nil {
+				c.AbortWithStatusJSON(http.StatusUnauthorized, err.Error())
+				return
+			}
+
+			if xPub == appConfig.Authentication.AdminKey {
+				c.Set(ParamAdminRequest, true)
+			}
+
+			c.Set(ParamXPubKey, xPub)
+
+		} else if authAccessKey != "" {
+			xPubOrAccessKey = authAccessKey
```

Comment:
Ineffectual assignment to `xPubOrAccessKey`

Reply:
Done